### PR TITLE
feat: implement VotingStreakNFTModule for Issue #24

### PR DESCRIPTION
**Pull request by PasqualePaoloSilvenni**
This PR introduces the VotingStreakNFTModule, representing a major architectural pivot from the previously proposed VotingStreakNFTStrategy. Based on the latest feedback, the streak logic has been entirely migrated from the Distribution domain (which targets recipients) to the Voting domain (which conceptually targets active voters).

### Key Features & Updates

- **Architecture** (Decorator Pattern): Implemented VotingStreakNFTModule by inheriting from the audited BasisPointsVotingModule. The core _processVote() hook is overridden to update user streaks in real-time. All custom state is safely isolated using EIP-7201 Namespaced Storage to guarantee 100% proxy-safety and prevent storage collisions.

- **Streak & Re-cast Logic**: Users must participate in consecutive cycles; missing a cycle resets their streak to 1. The module natively supports the parent's re-cast feature: voting multiple times within the same cycle triggers an early return (if (lastVoteCycle == currentCycle)), strictly preventing streak inflation/farming. A Crowdstake NFT is minted on the 10th consecutive vote.

- **Fail-Fast Minting Logic**: The module now assumes the external minting function is reliable: if an error occurs, the transaction simply reverts immediately. This makes the contract much more secure, readable, and cheaper to execute.

- **Testing Suite** (Harness Pattern): Deleted the obsolete strategy tests and completely rewrote the Foundry test suite (VotingStreakNFTModule.t.sol). Introduced a Test Harness to cleanly bypass the EIP-712 signature overhead. The suite now includes comprehensive AAA test cases covering core logic, streak resetting, and properly asserting the reverts of the new fail-fast architecture.

(Note: All legacy files related to the previous Distribution Strategy approach have been completely removed to keep the codebase clean).

### CODE REVIEW

File: src/implementation/strategies/VotingStreakNFTStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        uint256 cycle = currentCycle;
+        uint256 length = users.length;
+
+        for (uint256 i = 0; i < length; ) {
+            _processUser(users[i], cycle);
+            unchecked {
+                ++i;
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (RonTuretzky) [reply src/implementation/strategies/VotingStreakNFTStrategy.sol]: ignore this , its tracked in #85

File: src/implementation/strategies/VotingStreakNFTStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        if (_nftContract == address(0)) revert ZeroAddress();
+
+        VotingStreakNFTStrategyStorage storage $ = _getVotingStreakNFTStrategyStorage();
+        $.nftContract = IBreadkitNFT(_nftContract);
+        emit NFTContractUpdated(address(0), _nftContract);
+    }
+
+    // ============ Admin ============
+
+    /// @notice Updates the Breadkit NFT contract address
```
Comment (RonTuretzky) [inline src/implementation/strategies/VotingStreakNFTStrategy.sol]:
```suggestion
    /// @notice Updates the NFT contract address
```

**Comment by PasqualePaoloSilvenni** [comment]
Hey,
I've just pushed a major refactoring based on the latest feedback.
Here is a quick rundown of the architectural pivot:

- **Decorator Pattern:** I completely scrapped the old `DistributionStrategy` and implemented `VotingStreakNFTModule`, which cleanly extends the core `BasisPointsVotingModule`. It now hooks directly into `_processVote()` for real-time streak updates without modifying the audited core logic.
- **Test Harness Pattern:** Completely rewrote the Foundry test suite. I implemented a Test Harness to cleanly bypass the EIP-712 signature overhead, allowing for deep isolation testing. We now have **15 comprehensive AAA test cases** covering Core logic, Resiliency, and Edge Cases (all passing).
- **Cleanup:** Removed all legacy strategy files and obsolete mocks.

I've also updated the main PR description with the full technical documentation detailing the new architecture.

File: src/implementation/strategies/VotingStreakNFTStrategy.sol [hunk 3]
```diff
@@ -20,15 +20,30 @@ contract VotingStreakNFTStrategy is AbstractDistributionStrategy {
...
-
-    /// @notice Breadkit NFT contract used for streak rewards
-    IBreadkitNFT public nftContract;
+    /// @custom:storage-location erc7201:crowdstake.storage.VotingStreakNFTStrategy
+    struct VotingStreakNFTStrategyStorage {
+        /// @notice Current protocol cycle index
+        /// @dev Incremented on every execution path
+        uint256 currentCycle;
+        /// @notice Breadkit NFT contract used for streak rewards
+        IBreadkitNFT nftContract;
```
Comment (RonTuretzky) [inline src/implementation/strategies/VotingStreakNFTStrategy.sol]: please rename to crowdstake instead of breadkit

File: src/implementation/strategies/VotingStreakNFTStrategy.sol [hunk 4]
```diff
@@ -119,14 +154,16 @@ contract VotingStreakNFTStrategy is AbstractDistributionStrategy {
 
     /// @dev Increments cycle counter once per execution
     function _incrementCycle() internal {
```
Comment (RonTuretzky) [inline src/implementation/strategies/VotingStreakNFTStrategy.sol]: why does this contract have a cycle counter if there's already a cycle manager?

File: block
Comment (RonTuretzky) [inline block:1]: what is this?

**Review by RonTuretzky — COMMENTED** [review]
I dont see the hook for user voting here to determine their streak

File: src/implementation/strategies/VotingStreakNFTStrategy.sol (hunk 4 above)
Comment (PasqualePaoloSilvenni) [reply src/implementation/strategies/VotingStreakNFTStrategy.sol]: This entire file/logic was actually deprecated and removed in my latest push.

File: src/implementation/strategies/VotingStreakNFTStrategy.sol (hunk 3 above)
Comment (PasqualePaoloSilvenni) [reply src/implementation/strategies/VotingStreakNFTStrategy.sol]: Done

File: src/implementation/strategies/VotingStreakNFTStrategy.sol (hunk 2 above)
Comment (PasqualePaoloSilvenni) [reply src/implementation/strategies/VotingStreakNFTStrategy.sol]: Done

File: block
Comment (PasqualePaoloSilvenni) [reply block:1]: That was an unintended junk file generated locally by a stray console output. It shouldn't have been tracked by Git. I've deleted it in the latest push

**Comment by PasqualePaoloSilvenni** [comment]
> I dont see the hook for user voting here to determine their streak

In my latest push, I created the new VotingStreakNFTModule which safely overrides the _processVote() hook from the core voting module.

File: test/mocks/MockCycleModule.sol [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        return 100;
+    }
+
+    function updateCycleLength(uint256) external {}
+
+    function lastCycleStartBlock() external view returns (uint256) {
+        return block.number;
+    }
+
+    function cycleLength() external view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (PasqualePaoloSilvenni) [reply test/mocks/MockCycleModule.sol]: Done

File: test/VotingStreakNFTModule.t.sol [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+        (uint256 finalStreak2, , ) = harness.userActivity(user2);
+        assertEq(finalStreak1, 2, "user1 should maintain streak of 2");
+        assertEq(finalStreak2, 1, "user2 should have streak of 1");
+    }
+
+    /// @notice Test: Streak and NFT minting with zero address NFT contract
+    /// @dev Verifies graceful handling when NFT contract is not set
+    function test_GracefulHandlingWithoutNFTContract() public {
+        // Note: This test assumes NFT contract can be set to address(0) after initialization
+        // Arrange
```
[the remark this answers is not in the document]
Comment (PasqualePaoloSilvenni) [reply test/VotingStreakNFTModule.t.sol]: Spot on. That was an orphaned test template. I've completely removed test_GracefulHandlingWithoutNFTContract in the latest commit so it doesn't inflate the passing test metrics. Thanks!

File: src/implementation/VotingStreakNFTModule.sol [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+                activity.mintPending = false;
+            }
+            // If mint failed again, mintPending stays true for next retry opportunity
+        }
+    }
+
+    /// @dev Attempts to mint an NFT for a user with graceful error handling
+    /// @param user The user address to mint for
+    /// @return success True if mint succeeded, false otherwise
+    function _attemptMint(address user) internal returns (bool success) {
```
Comment (RonTuretzky) [inline src/implementation/VotingStreakNFTModule.sol]: This function invokes another storage pull , and I don’t see the value in separating this functionality out to an internal function when it is not used again anywhere else

File: src/interfaces/ICrowdstakeNFT.sol [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+/// @dev Extends IERC721 by adding mint and burn functions used by the Crowdstake distribution strategy and other external consumers.
+interface ICrowdstakeNFT is IERC721 {
+    /// @notice Mints a new NFT to the specified address.
+    /// @param to The address that will receive the NFT.
+    /// @return tokenId The unique ID of the newly minted NFT.
+    function mint(address to) external returns (uint256);
+
+    /// @notice Burns a specific NFT.
+    /// @param tokenId The ID of the NFT to burn.
+    function burn(uint256 tokenId) external;
```
Comment (RonTuretzky) [inline src/interfaces/ICrowdstakeNFT.sol]: Unclear why burn functionality is introduced if it is unused

File: foundry.lock
Comment (RonTuretzky) [inline foundry.lock:1]: Introducing a new file to the project , out of scope

**Review by RonTuretzky — CHANGES_REQUESTED** [review]

File: src/interfaces/ICrowdstakeNFT.sol (hunk 8 above)
Comment (PasqualePaoloSilvenni) [reply src/interfaces/ICrowdstakeNFT.sol]: I originally included burn() because it was explicitly mentioned in the initial Issue specs ('Optionally burns NFTs for users who no longer meet holding criteria'), but I simply forgot to deprecate it after our pivot to the voting streaks. I've removed it now!

File: src/implementation/VotingStreakNFTModule.sol (hunk 7 above)
Comment (PasqualePaoloSilvenni) [reply src/implementation/VotingStreakNFTModule.sol]: I initially separated it for readability, but you're absolutely right about the redundant SLOAD cost. I've inlined this logic directly into the main execution flow to eliminate the extra storage read and the unnecessary abstraction. Pushed the fix!

File: foundry.lock
Comment (PasqualePaoloSilvenni) [reply foundry.lock:1]: My apologies! I've run git rm --cached foundry.lock to untrack it and keep this PR strictly in scope.

File: src/implementation/VotingStreakNFTModule.sol [hunk 9]
```diff
@@ -200,45 +200,45 @@ contract VotingStreakNFTModule is BasisPointsVotingModule {
...
+            } else {
+                emit NFTMintFailed(user, bytes("NFT contract not set"));
+            }
+            if (mintSuccess) {
                 activity.mintPending = false;
             }
             // If mint failed, mintPending stays true for retry in next cycle when streak > 10
         }
         // Condition 2: Mint is pending and streak continues beyond 10
         else if (activity.mintPending && activity.streak > 10) {
```
Comment (RonTuretzky) [inline src/implementation/VotingStreakNFTModule.sol]: Actually I don’t understand the scenario this can even happen in ? Why wouldn’t they just get it on the 10th? In what situation are they voting 10 times and not getting minted?

Comment (PasqualePaoloSilvenni) [reply src/implementation/VotingStreakNFTModule.sol]: They absolutely should get it on the 10th vote. This block is strictly a fallback/retry mechanism for our Graceful Degradation implementation.
Here is the exact edge-case scenario where this happens:

1. User casts their 10th consecutive vote.
2. The try/catch block attempts nftContract.mint(), but the external NFT contract temporarily fails (e.g., it's paused, out-of-gas in the subcall, etc.).
3. Instead of reverting the entire transaction and blocking their governance vote, we gracefully catch the error and set mintPending = true.
4. In the next cycle, the user casts their 11th consecutive vote (streak > 10).
5. This else if block catches that they are owed an NFT (mintPending == true) and automatically retries the minting process!

Comment (RonTuretzky) [reply src/implementation/VotingStreakNFTModule.sol]: My friend can you please your your brain here , the ai is not understanding what I’m saying haha

Comment (RonTuretzky) [reply src/implementation/VotingStreakNFTModule.sol]: We can assume the nft mint function will work , we should not support this edge case

Comment (PasqualePaoloSilvenni) [reply src/implementation/VotingStreakNFTModule.sol]: Haha, understood! My bad, I’ll proceed to clean the minting logic, leaving only the ten streak case

**Review by RonTuretzky — APPROVED** [review]

**Comment by RonTuretzky** [comment]
@bagelface  hey LGTM if you have capacity to give this a quick review

**Review by Tranquil-Flow — COMMENTED** [review]
## Overall

Good approach with the decorator pattern for streak tracking. The core streak counting logic is correct. However, there are 3 functional issues and a few smaller items to address before merge.

---

### 🔴 `streak == 10` only fires once — never again

```solidity
if (activity.streak == 10) {
    $.nftContract.mint(user);
}
```

After minting at streak=10, the streak becomes 11, 12, 13… and the condition is never true again. The code comment on line 184 says "and multiples thereafter" but the implementation doesn't match. If the intent is every 10th streak:

```solidity
if (activity.streak > 0 && activity.streak % 10 == 0) {
    $.nftContract.mint(user);
}
```

Also: the PR title says "NFT awarding **and burning**" but burn is not implemented. Either implement it or update the title.

---

### 🔴 Minted token IDs are discarded

```solidity
$.nftContract.mint(user);  // returns uint256 tokenId — discarded
```

The module has no record of which token ID was issued. If burn logic is added later, there's no on-chain way to know which token to burn without extra infrastructure (e.g., `ERC721Enumerable`).

**Fix:** Store the token ID:
```solidity
uint256 tokenId = $.nftContract.mint(user);
$.tokenIds[user].push(tokenId);
```

---

### 🟡 No reentrancy protection on vote functions

The mock NFT uses `_safeMint`, which triggers `onERC721Received()` on contract recipients. A malicious contract could re-enter `castVoteWithSignature()` during that callback with a different pre-signed nonce. While the nonce check prevents same-nonce replay, a second valid signature could manipulate voting power mid-execution.

Whether this is exploitable depends on the production NFT implementation — if it uses `_mint` instead of `_safeMint`, there's no callback. But defensive practice says add `nonReentrant`:

```solidity
function castVoteWithSignature(...) external nonReentrant ...
```

---

### 🟡 Minor items

- **No event on mint:** When streak hits 10 and mint fires, only `StreakUpdated` is emitted — no dedicated `NFTMinted(user, tokenId)` event. Makes off-chain indexing harder.
- **Misleading comment:** Line 184 says "and multiples thereafter" but code only checks `== 10`.

---

### ✅ What's working well

- Streak counting logic (consecutive cycle tracking, reset on missed cycle) is correct
- The `_getBasisPointsVotingModuleStorage()` visibility fix is necessary and correct
- Test coverage for streak counting, mint at 10, recast protection, and multi-user isolation is solid
- ICrowdstakeNFT interface design is clean
- Decorator pattern is the right architecture choice — keeps streak logic separate from core voting

**Comment by PasqualePaoloSilvenni** [comment]
The "burning" mention actually comes from the original Issue #24 title, not the PR. I dropped the burn logic for now since the architecture pivoted to streak-based voting. I don't have the permissions to update the issue's title.
Just pushed a commit addressing all the feedback:
- Used % 10 modulo for multiple streak milestones
- Tracked tokenIds in EIP-7201 storage (and exposed them via a public getter)
- Added NFTMinted event for off-chain indexing
- Secured voting functions with ReentrancyGuardUpgradeable
- Updated the test suite according to the changes

Let me know if we're good to merge!

**Comment by RonTuretzky** [comment]
@PasqualePaoloSilvenni you implemented an AI comment ... please revert [07cd543](https://github.com/BreadchainCoop/crowdstake.fun/pull/82/commits/07cd5435f19c97ef5084387ce0528b190eb30cb7)
[e3c072e](https://github.com/BreadchainCoop/crowdstake.fun/pull/82/commits/e3c072ecfe51b6174ecb4aa7023e96c4a16f76d1)
[e56e8b8](https://github.com/BreadchainCoop/crowdstake.fun/pull/82/commits/e56e8b8f37b19763ba7fdf9fee271ae31302ffbe)
[d49d842](https://github.com/BreadchainCoop/crowdstake.fun/pull/82/commits/d49d842b68af8184b73f6ed21604e6baaf7432f8)

**Comment by PasqualePaoloSilvenni** [comment]
I genuinely thought it was human! I've just reverted the commits and cleaned up the branch. The PR is now back to its original state.
